ff/src/fields/sqrt.rs

```diff
@@ -65,14 +65,55 @@ impl LegendreSymbol {
 /// the corresponding condition.
 #[non_exhaustive]
 pub enum SqrtPrecomputation<F: crate::Field> {
-    // Tonelli-Shanks algorithm works for all elements, no matter what the modulus is.
+    /// https://eprint.iacr.org/2012/685.pdf (page 12, algorithm 5).
+    /// Tonelli-Shanks algorithm works for all elements, no matter what the modulus is.
+    /// With _q_ as field order, _p_ as characteristic, and _m_ as extension degree:
+    /// * First factor _q - 1 = 2^s t_ where _t_ is odd.
+    /// * `two_adicity` - _s_.
+    /// * `quadratic_nonresidue_to_trace` - _c^t_, with random _c_ such that _c^2^(s - 1) = 1_.
+    /// * `trace_of_modulus_minus_one_div_two` - _(t - 1)/2_.
     TonelliShanks {
         two_adicity: u32,
         quadratic_nonresidue_to_trace: F,
         trace_of_modulus_minus_one_div_two: &'static [u64],
     },
-    /// To be used when the modulus is 3 mod 4.
-    Case3Mod4 {
+    /// https://eprint.iacr.org/2012/685.pdf (page 9, algorithm 2).
+    /// With _q_ as field order, _p_ as characteristic, and _m_ as extension degree:
+    /// * `char_minus_three_div_four` - _(p - 3)/4_.
+    /// * `deg_minus_three_div_two_plus_one` - _(m - 3)/2 + 1_.
+    ShanksCase3Mod4 {
+        char_minus_three_div_four: &'static [u64],
+        deg_minus_three_div_two_plus_one: usize,
```

Comment:
We don't need to have this as a separate field, since `F::extension_degree()` exists.

Reply:
I was thinking this would save some computation but sure